File: next-fx/src/main/java/com/rememberber/mootool/nextfx/domain/encode/EncodeEngine.java
package com.rememberber.mootool.nextfx.domain.encode;

import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.Charset;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Encode/decode algorithms aligned with Electron {@code encodeTools.ts}.
 */
public final class EncodeEngine {

    private static final Pattern UNICODE_ESCAPE = Pattern.compile("\\\\u([\\da-fA-F]{4})");
    private static final Pattern HEX_PAIR = Pattern.compile("[\\da-fA-F]{2}");
    private static final Pattern TOKEN_SPLIT = Pattern.compile("[\\s,;]+");
    private static final Charset GB2312 = Charset.forName("GB2312");

    public enum Tab {
        UNICODE,
        URL,
        HEX,
        ASCII;

        public String id() {
            return name().toLowerCase(Locale.ROOT);
        }

        public static Tab fromId(String id) {
            if (id == null) {
                return UNICODE;
            }
            return switch (id.toLowerCase(Locale.ROOT)) {
                case "url" -> URL;
                case "hex" -> HEX;
                case "ascii" -> ASCII;
                default -> UNICODE;
            };
        }
    }

    public enum Direction {
        FORWARD,
        REVERSE
    }

    public enum UrlCharset {
        UTF_8("utf-8", StandardCharsets.UTF_8),
        GB2312("gb2312", EncodeEngine.GB2312);

        private final String id;
        private final Charset charset;

        UrlCharset(String id, Charset charset) {
            this.id = id;
            this.charset = charset;
        }

        public String id() {
            return id;
        }

        public Charset charset() {
            return charset;
        }

        public static UrlCharset fromId(String id) {
            if (id == null) {
                return UTF_8;
            }
            return "gb2312".equalsIgnoreCase(id) ? GB2312 : UTF_8;
        }
    }

    public enum AsciiFormat {
        DECIMAL,
        HEX;

        public static AsciiFormat fromId(String id) {
            return "hex".equalsIgnoreCase(id) ? HEX : DECIMAL;
        }

        public String id() {
            return name().toLowerCase(Locale.ROOT);
        }
    }

    private EncodeEngine() {
    }

    public static String convert(Tab tab, Direction direction, String value, UrlCharset charset, AsciiFormat asciiFormat) {
        String input = value == null ? "" : value;
        return switch (tab) {
            case UNICODE -> direction == Direction.FORWARD ? toUnicode(input) : fromUnicode(input);
            case URL -> direction == Direction.FORWARD ? urlEncode(input, charset) : urlDecode(input, charset);
            case HEX -> direction == Direction.FORWARD ? textToHex(input) : hexToText(input);
            case ASCII -> direction == Direction.FORWARD ? textToAscii(input, asciiFormat) : asciiToText(input);
        };
    }

    public static String toUnicode(String value) {
        StringBuilder out = new StringBuilder();
        value.codePoints().forEach(codePoint -> {
            if (codePoint <= 0x7f) {
                out.appendCodePoint(codePoint);
            } else if (codePoint <= 0xffff) {
                out.append("\\u").append(String.format("%04x", codePoint));
            } else {
                int offset = codePoint - 0x10000;
                int high = 0xd800 + (offset >> 10);
                int low = 0xdc00 + (offset & 0x3ff);
                out.append("\\u").append(Integer.toHexString(high));
                out.append("\\u").append(Integer.toHexString(low));
            }
        });
        return out.toString();
    }

    public static String fromUnicode(String value) {
        Matcher matcher = UNICODE_ESCAPE.matcher(value);
        StringBuilder out = new StringBuilder();
        while (matcher.find()) {
            matcher.appendReplacement(out, Matcher.quoteReplacement(
                    String.valueOf((char) Integer.parseInt(matcher.group(1), 16))));
        }
        matcher.appendTail(out);
        return out.toString();
    }

    public static String urlEncode(String value, UrlCharset charset) {
        byte[] bytes = value.getBytes(charset.charset());
        StringBuilder out = new StringBuilder(bytes.length * 3);
        for (byte raw : bytes) {
            int unsigned = raw & 0xff;
            if (isUnreserved(unsigned)) {
                out.append((char) unsigned);
            } else {
                out.append('%');
                out.append(String.format("%02X", unsigned));
            }
        }
        return out.toString();
    }

    public static String urlDecode(String value, UrlCharset charset) {
        String source = value.replace('+', ' ');
        List<Byte> bytes = new ArrayList<>();
        for (int index = 0; index < source.length(); ) {
            if (source.charAt(index) == '%'
                    && index + 2 < source.length()
                    && HEX_PAIR.matcher(source.substring(index + 1, index + 3)).matches()) {
                bytes.add((byte) Integer.parseInt(source.substring(index + 1, index + 3), 16));
                index += 3;
            } else {
                for (byte raw : String.valueOf(source.charAt(index)).getBytes(charset.charset())) {
                    bytes.add(raw);
                }
                index += 1;
            }
        }
        byte[] array = new byte[bytes.size()];
        for (int i = 0; i < bytes.size(); i++) {
            array[i] = bytes.get(i);
        }
        return new String(array, charset.charset());
    }

    public static String textToHex(String value) {
        byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
        StringBuilder out = new StringBuilder(bytes.length * 2);
        for (byte raw : bytes) {
            out.append(String.format("%02x", raw & 0xff));
        }
        return out.toString();
    }

    public static String hexToText(String value) {
        String normalized = value.replaceAll("[\\s:_-]+", "");
        if (normalized.isEmpty() || normalized.length() % 2 != 0 || !normalized.matches("[\\da-fA-F]+")) {
            throw new EncodeException("encode.error.invalidHex", "");
        }
        byte[] bytes = new byte[normalized.length() / 2];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = (byte) Integer.parseInt(normalized.substring(i * 2, i * 2 + 2), 16);
        }
        CharsetDecoder decoder = StandardCharsets.UTF_8.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT);
        try {
            return decoder.decode(ByteBuffer.wrap(bytes)).toString();
        } catch (CharacterCodingException exception) {
            throw new EncodeException("encode.error.invalidHex", exception.getMessage());
        }
    }

    public static String textToAscii(String value, AsciiFormat format) {
        StringBuilder out = new StringBuilder();
        value.codePoints().forEach(code -> {
            if (!out.isEmpty()) {
                out.append(' ');
            }
            out.append(format == AsciiFormat.HEX
                    ? Integer.toHexString(code).toUpperCase(Locale.ROOT)
                    : Integer.toString(code));
        });
        return out.toString();
    }

    public static String asciiToText(String value) {
        String trimmed = value.trim();
        if (trimmed.isEmpty()) {
            return "";
        }
        StringBuilder out = new StringBuilder();
        for (String part : TOKEN_SPLIT.split(trimmed)) {
            if (part.isEmpty()) {
                continue;
            }
            boolean hex = part.regionMatches(true, 0, "0x", 0, 2) || part.matches("(?i).*[a-f].*");
            String normalized = part.replaceFirst("(?i)^0x", "");
            int codePoint;
            try {
                codePoint = Integer.parseInt(normalized, hex ? 16 : 10);
            } catch (NumberFormatException exception) {
                throw new EncodeException("encode.error.invalidCodePoint", part);
            }
            if (codePoint < 0 || codePoint > 0x10ffff) {
                throw new EncodeException("encode.error.invalidCodePoint", part);
            }
            out.appendCodePoint(codePoint);
        }
        return out.toString();
    }

    private static boolean isUnreserved(int unsignedByte) {
        return (unsignedByte >= 0x41 && unsignedByte <= 0x5a)
                || (unsignedByte >= 0x61 && unsignedByte <= 0x7a)
                || (unsignedByte >= 0x30 && unsignedByte <= 0x39)
                || unsignedByte == 0x2d
                || unsignedByte == 0x2e
                || unsignedByte == 0x5f
                || unsignedByte == 0x7e;
    }
}

File: next-compose/composeApp/src/desktopMain/kotlin/com/rememberber/mootool/next/compose/domain/EncodeEngine.kt
package com.rememberber.mootool.next.compose.domain

import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.Charset
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets

enum class EncodeTab { Unicode, Url, Hex, Ascii }

enum class UrlCharset { Utf8, Gb2312 }

enum class AsciiFormat { Decimal, Hex }

class EncodeException(val code: String, message: String) : RuntimeException(message)

object EncodeEngine {
    private val unicodeEscape = Regex("""\\u([0-9a-fA-F]{4})""")
    private val hexPair = Regex("..")
    private val asciiParts = Regex("""[\s,;]+""")
    private val gb2312: Charset = Charset.forName("GB2312")

    fun convert(
        tab: EncodeTab,
        forward: Boolean,
        value: String,
        charset: UrlCharset,
        asciiFormat: AsciiFormat
    ): String = when (tab) {
        EncodeTab.Unicode -> if (forward) toUnicode(value) else fromUnicode(value)
        EncodeTab.Url -> if (forward) urlEncode(value, charset) else urlDecode(value, charset)
        EncodeTab.Hex -> if (forward) textToHex(value) else hexToText(value)
        EncodeTab.Ascii -> if (forward) textToAscii(value, asciiFormat) else asciiToText(value)
    }

    fun toUnicode(value: String): String = buildString {
        var index = 0
        while (index < value.length) {
            val codePoint = value.codePointAt(index)
            index += Character.charCount(codePoint)
            when {
                codePoint <= 0x7f -> appendCodePoint(codePoint)
                codePoint <= 0xffff -> append("\\u").append(codePoint.toString(16).padStart(4, '0'))
                else -> {
                    val offset = codePoint - 0x10000
                    val high = 0xd800 + (offset shr 10)
                    val low = 0xdc00 + (offset and 0x3ff)
                    append("\\u").append(high.toString(16))
                    append("\\u").append(low.toString(16))
                }
            }
        }
    }

    fun fromUnicode(value: String): String = unicodeEscape.replace(value) { match ->
        Char(match.groupValues[1].toInt(16)).toString()
    }

    fun urlEncode(value: String, charset: UrlCharset): String {
        val bytes = encodeText(value, charsetOf(charset))
        return buildString(bytes.size * 3) {
            for (byte in bytes) {
                val unsigned = byte.toInt() and 0xff
                if (isUnreserved(unsigned)) append(unsigned.toChar())
                else append('%').append(unsigned.toString(16).uppercase().padStart(2, '0'))
            }
        }
    }

    fun urlDecode(value: String, charset: UrlCharset): String {
        val encoding = charsetOf(charset)
        val source = value.replace('+', ' ')
        val bytes = ArrayList<Byte>(source.length)
        var index = 0
        while (index < source.length) {
            val current = source[index]
            if (current == '%' && index + 2 < source.length && isHexPair(source, index + 1)) {
                bytes += source.substring(index + 1, index + 3).toInt(16).toByte()
                index += 3
            } else {
                encodeText(current.toString(), encoding).forEach { bytes += it }
                index += 1
            }
        }
        return decodeText(bytes.toByteArray(), encoding)
    }

    fun textToHex(value: String): String =
        value.toByteArray(StandardCharsets.UTF_8).joinToString("") { byte ->
            (byte.toInt() and 0xff).toString(16).padStart(2, '0')
        }

    fun hexToText(value: String): String {
        val normalized = value.replace(Regex("""[\s:_-]+"""), "")
        if (normalized.isEmpty() || normalized.length % 2 != 0 || !normalized.matches(Regex("""^[0-9a-fA-F]+$"""))) {
            throw EncodeException("invalid-hex", "invalid-hex")
        }
        val bytes = hexPair.findAll(normalized).map { it.value.toInt(16).toByte() }.toList().toByteArray()
        return decodeText(bytes, StandardCharsets.UTF_8)
    }

    fun textToAscii(value: String, format: AsciiFormat): String = buildString {
        var index = 0
        var first = true
        while (index < value.length) {
            val codePoint = value.codePointAt(index)
            index += Character.charCount(codePoint)
            if (!first) append(' ')
            first = false
            append(if (format == AsciiFormat.Hex) codePoint.toString(16).uppercase() else codePoint.toString())
        }
    }

    fun asciiToText(value: String): String {
        val trimmed = value.trim()
        if (trimmed.isEmpty()) return ""
        return trimmed.split(asciiParts).joinToString("") { part ->
            val hex = part.startsWith("0x", ignoreCase = true) || part.contains(Regex("[a-fA-F]"))
            val normalized = part.replace(Regex("^0x", RegexOption.IGNORE_CASE), "")
            val codePoint = normalized.toIntOrNull(if (hex) 16 else 10)
                ?: throw EncodeException("invalid-code-point", part)
            if (codePoint < 0 || codePoint > 0x10ffff) throw EncodeException("invalid-code-point", part)
            if (codePoint in 0xd800..0xdfff) Char(codePoint).toString()
            else String(Character.toChars(codePoint))
        }
    }

    private fun charsetOf(charset: UrlCharset): Charset = when (charset) {
        UrlCharset.Utf8 -> StandardCharsets.UTF_8
        UrlCharset.Gb2312 -> gb2312
    }

    private fun encodeText(value: String, charset: Charset): ByteArray {
        val encoder = charset.newEncoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
        return try {
            val buffer = encoder.encode(CharBuffer.wrap(value))
            val bytes = ByteArray(buffer.remaining())
            buffer.get(bytes)
            bytes
        } catch (_: CharacterCodingException) {
            throw EncodeException("unmappable", "unmappable")
        }
    }

    private fun decodeText(bytes: ByteArray, charset: Charset): String {
        val decoder = charset.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
        return try {
            decoder.decode(ByteBuffer.wrap(bytes)).toString()
        } catch (_: CharacterCodingException) {
            throw EncodeException("invalid-bytes", "invalid-bytes")
        }
    }

    private fun isUnreserved(byte: Int): Boolean {
        return (byte in 0x41..0x5a) || (byte in 0x61..0x7a) || (byte in 0x30..0x39) ||
            byte == 0x2d || byte == 0x2e || byte == 0x5f || byte == 0x7e
    }

    private fun isHexPair(source: String, start: Int): Boolean {
        val first = source[start]
        val second = source[start + 1]
        return first.isHex() && second.isHex()
    }

    private fun Char.isHex(): Boolean = this in '0'..'9' || this in 'a'..'f' || this in 'A'..'F'
}
